ArduPlane/ArduPlane.cpp

```diff
@@ -968,6 +968,86 @@ void Plane::update_flight_stage(void)
     airspeed.set_EAS2TAS(barometer.get_EAS2TAS());
 }
 
+/*
+*  ArduSoar support function
+*
+*  Peter Braswell and Samuel Tabor
+*/
+void Plane::update_soaring() {
+    
+    if (!soaring_controller.is_active()) {
+        return;
+    }
+    
+    soaring_controller.update_vario();
+
+    // Check for throttle suppression change.
+    switch (control_mode)
+    {
+    case AUTO:
+        soaring_controller.suppress_throttle();
+        break;
+    case FLY_BY_WIRE_B:
+        if (!soaring_controller.suppress_throttle()) {
+            ;
+            set_mode(RTL, MODE_REASON_FBW_B_WITH_MOTOR_RUNNING);
+        }
+        break;
+    case LOITER:
+        // Do nothing. We will switch back to auto/rtl before enabling throttle.
+        break;
+    default:
+        // This does not affect the throttle since suppressed is only checked in the above three modes. 
+        // It ensures that the soaring always starts with throttle suppressed though.
+        soaring_controller.set_throttle_suppressed(true);
+        break;
+    }
+    // Nothing to do if we are in powered flight
+    if (!soaring_controller.get_throttle_suppressed() && aparm.throttle_max>0) {
+        return;
+    }
+
+    switch (control_mode)
+    {
+    case AUTO:
+    case FLY_BY_WIRE_B:
+        // Test for switch into thermalling mode
+        soaring_controller.update_cruising();
+
+        if (soaring_controller.check_thermal_criteria()) {
+            hal.console->printf("Thermal detected, entering loiter\n");
+            set_mode(LOITER, MODE_REASON_THERMAL_DETECTED);
```

Comment:
Not necessarily -- as you pointed it, it depends on your speed, but also on the actual size, strength, and position of a given thermal. Usually, the center of the updraft is usually somewhere off your course. Setting the centerpoint either directly ahead or behind you is a rather primitive heuristic, but it's a start. Taking velocity into account won't improve it much. Ideally, we want to consider approaches such as run the thermal EKF continuously and triggering a transition to thermalling/loiter when your confidence of thermal center position (e.g., as measured by your estimate variance) becomes sufficiently small. For now though, I'd like to check it in as is to give others a chance to explore ideas in this space.